cranelift/codegen/src/isa/x64/encoding/evex.rs

```diff
@@ -151,10 +178,44 @@ impl EvexInstruction {
     /// - finally, the ModR/M byte.
     ///
     /// Eventually this method should support encodings of more than just the reg-reg addressing mode (TODO).
-    pub fn encode<CS: ByteSink + ?Sized>(&self, sink: &mut CS) {
+    pub fn encode(&self, sink: &mut MachBuffer<Inst>) {
+        if let RegisterOrAmode::Amode(amode) = &self.rm {
+            if amode.can_trap() {
+                sink.add_trap(TrapCode::HeapOutOfBounds);
+            }
+        }
         sink.put4(self.bits);
         sink.put1(self.opcode);
-        sink.put1(encode_modrm(3, self.reg.0 & 7, self.rm.0 & 7));
+
+        match &self.rm {
+            RegisterOrAmode::Register(reg) => {
+                let rm: u8 = (*reg).into();
+                sink.put1(rex::encode_modrm(3, self.reg.0 & 7, rm & 7));
+            }
+            RegisterOrAmode::Amode(amode) => {
+                // NB: the `scaling` factor must be passed to
+                // `emit_modrm_sib_disp` because EVEX instructions always use a
+                // scaling factor. Section 2.7.5 in the Intel manual has a lot
+                // of words about this I don't fully understand yet. That being
+                // said for the V128 case the factor is 16 (or so I'm pretty
+                // sure). Assert that that's the "length" of this instruction
+                // (aka the encoding of the default `EvexContext`). If
+                // this assertion trips then the scaling factor probably needs
+                // to be stored on `self` and reassigned whenever the `length`
+                // method is called.
+                let b128 = EvexContext::Other {
+                    length: EvexVectorLength::V128,
+                };
+                let bits = (self.bits >> Self::LL.start()) & ((1 << Self::LL.len()) - 1);
+                assert_eq!(bits, u32::from(b128.bits()));
+                let scaling = 16;
```

Comment:
I think I'm convinced from the tests that this scaling factor is correct for `vpabs` but I'm not sure this is set up well for the future (or even all of the possible EVEX-encoded instructions that can currently be emitted). Section 2.7.7 talks about the kinds of instructions that do not have embedded broadcast (single scalar result, explicit broadcast in opcode, pure load/store) &mdash; these instructions all have `scaling != 16` for `VL = 128`. I think this means that if any future or current instructions are of this kind, we could get a miscompilation (see the expected scaling factors in table 2-35). I wish there was a way to throw an error of some kind here instead. (The same kind of logic applies for the instructions that _are_ affected by embedded broadcast: instructions could expect scalings of 8 and 4 unless we do more checks, e.g., on `EVEX.b` and `EVEX.W`).

My argument here is "let's prevent errors" when someone tries to add a new EVEX-encoded instruction in the future. I looked into adding a new function to `Avx512Opcode` to signal which tuple type each opcode has but that might not totally work: `vpopcnt`, e.g., can be emitted as both "Full Mem" and "Full" (we use the "Full Mem" one, apparently!). `vcvtudq2ps` uses "Full", which only uses `scaling = 16` when `EVEX.b = 0` and `EVEX.W = 1`. Maybe `EvexInstruction` needs to gain a `tuple` field which we force users to specify somehow in `emit.rs`. Not sure I have the best solution here but hopefully you get what I mean about protecting against future errors.

Reply:
Oh my read was that `EvexContext` was what controlled all these bits where whatever sets `EvexContext` guides the scaling, but it sounds like that's not right. I see now the listing on each instruction and its "TupleType"!

Looking at [vpopcnt](https://www.felixcloutier.com/x86/vpopcnt) if we have a separate opcode per-size (which I think we do today anyway) it looks like that means the tuple type is per-opcode still? If that's possible then I could add something like:

```rust
impl Avx512Opcode {
    fn tuple_type(&self) -> Avx512TupleType { /* ... */ }
}
```

which could get fed into the `EvexInstruction` encoder which would then futher handle all the `EVEX.{b,W}` cases? That way if more tuple types were added they would result in a non-exhaustive match and require handling?